# Spam Detector

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [2]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

I predict that the random forest model will perform better. It seems that numerous weak predictors would do a better job with the interrelated nature of multi-factor spam prediction than logistic regression which considers each factor more independently. 

## Split the Data into Training and Testing Sets

In [5]:
# Create the labels set `y` and features DataFrame `X`
y = data[['spam']]
X = data.drop(['spam'], axis=1)

In [6]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
y.value_counts()

spam
0       2788
1       1813
Name: count, dtype: int64

In [15]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [16]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance
scaler = StandardScaler()

In [17]:
# Fit the Standard Scaler with the training data
X_scaler = scaler.fit(X_train)

In [18]:
# Scale the training data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [19]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression

lr_model = LogisticRegression(solver='lbfgs', random_state=1)
lr_model.fit(X_train_scaled, y_train.values.ravel())
lr_train_score = lr_model.score(X_train_scaled, y_train.values.ravel())

print(f"Logistic Regression Training Score: {lr_train_score:.3f}")

Logistic Regression Training Score: 0.930


In [20]:
# Make and save testing predictions with the saved logistic regression model using the test data
lr_test_predictions = lr_model.predict(X_test_scaled)

# Review the predictions
lr_test_predictions

array([0, 0, 1, ..., 0, 0, 1])

In [21]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
lr_test_score = accuracy_score(y_test, lr_test_predictions)

print(f"Logistic Regression Test Score: {lr_test_score:.3f}")

Logistic Regression Test Score: 0.929


## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [22]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier

rf_model = RandomForestClassifier(n_estimators=128, random_state=1)
rf_model.fit(X_train_scaled, y_train.values.ravel())
rf_train_score = rf_model.score(X_train_scaled, y_train.values.ravel())

print(f"Random Forest Training Score: {rf_train_score:.3f}")

Random Forest Training Score: 1.000


In [23]:
# Make and save testing predictions with the saved logistic regression model using the test data
rf_test_predictions = rf_model.predict(X_test_scaled)

# Review the predictions
rf_test_predictions

array([1, 1, 1, ..., 1, 0, 1])

In [24]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
rf_test_score = accuracy_score(y_test, rf_test_predictions)

print(f"Random Forest Test Score: {rf_test_score:.3f}")

Random Forest Test Score: 0.967


## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

In [31]:
score_comp = pd.DataFrame([(lr_train_score, lr_test_score),
                           (rf_train_score, rf_test_score)],
                          columns=['Training Score', 'Test Score'],
                          index=['Logistic Regression', 'Random Forest']).round(4)
score_comp

Training Score  Test Score
Logistic Regression          0.9299      0.9288
Random Forest                0.9997      0.9670

## Conclusions

As predicted, the Random Forest model outperformed the Logistic Regression model in this spam classification task. 

Both models achieved high accuracy scores, with the Random Forest model achieving a test score of 96.7% and the Logistic Regression model achieving a test score of 92.9%. However, the Random Forest model had a higher training score of 99.9% compared to the Logistic Regression model's training score of 92.9%. This indicates that the Random Forest model was able to better capture the underlying patterns and relationships in the data, resulting in improved performance on unseen test data. 

It's interesting to not that the accuracy score for both the training and test sets in the logistic regression were near equivalent. This was not the case for the Random forest - here the testing accuracy was 3% lower than the training accuracy. It's possible that the random forest has overfit the data and does not generalize as well as the logistic regression. It may be worth continuing to monitior both models as new data is acquired. 

That said, both models exhibited stellar performance with this data set. In the subtle competition, the Random Forest model's ability to combine multiple decision trees and consider a larger number of features likely contributed to its superior performance in this case.

